doc/design/record_fetcher_proposal.md

```diff
@@ -111,31 +107,27 @@ manifest's information in the JWT payload. As an example, the payload could rese
       "label": "some pdf we'll have to ocr",
       "content_type": "application/pdf",
       "digest": "...",
-      "uri": "https://..."
     }, {
       "label": "party information",
       "content_type": "text/html",
       "digest": "...",
-      "uri": "https://..."
     }
   ]
 }	
 ```
 
-Determining the PDAP-controlled URIs of the documents will require uploading them first, and then constructing the
-manifest.
-
-### Garbage Collection
+### Record Bundles
 
-Since this design requires documents to be uploaded before writing out a manifest, it's possible that we'll upload docs
-and then fail (or neglect) to upload a manifest. This results in orphaned documents, which are just wasted storage. 
-To mitigate this, we'll construct a GC batch job that runs on a schedule and removes documents that are orphaned and
-have existed for longer than X time.
+Signed manifests and all related documents will be uploaded together as [MIME Multipart messages](
```

Comment:
Just reiterating that I'm not sure this is a concern if this all happens under our control.

Reply:
mTLS was the original plan, but that just gets us endpoint auth, so I ditched that in favor of the JWT bearer token (not to be confused with the signed manifest JWT) because it's simpler. That said, I wanted things to be tamper evident even while in PDAP-controlled storage, as I don't really trust PDAP-controlled storage to not be tampered with.

We're working on exposing government accountability data to journalists, activists, lawmakers, etc. in situations where it's difficult to collect records to begin with (and therefore difficult to verify our data), so tampering with our data could be highly valuable for some parties. We're a loose coalition of volunteers where no one knows anyone (some folks aren't even using their real names), so we can't *really* put a ton of trust in anyone, and I don't know who would have write access to the raw data.

Targeted tampering with records seems low effort and high reward, so I'm trying to make that tampering evident at a fundamental level. Thoughts?

Reply:
The idea of not trusting PDAP-controlled storage is a bit concerning to me. Hopefully, you mean you don't trust what is stored there, which is more reasonable. The question of where to draw lines of what is trustworthy is still there, though. If we can't trust what is running within our infrastructure, how can we trust the infrastructure managing JWT tokens and their authentication?

Reply:
> The idea of not trusting PDAP-controlled storage is a bit concerning to me. Hopefully, you mean you don't trust what is stored there, which is more reasonable.

Yeah, sorry about that. "Not trusting storage" has implications I didn't mean to convey. From an insider risk angle (or just misconfiguration) angle, I'm hesitant to assume that nothing in our storage has been tampered with. I think what I'm mostly shooting for is a model where we largely trust our storage and its contents, but are able to scalably verify they are trustworthy.

> EDIT: Just as an additional note, security is not really my wheelhouse.

My present work is security adjacent, but this is still largely not my wheelhouse either. 😄 I may pop a link to this in the infosec chat, since the discussions surrounding this has drifted into that territory. On the topic of things that aren't wheelhouses, I'm not well-versed in what the actual landscape looks like here (e.g., is there something in AWS that makes some aspect of this moot?), so there could be some assumptions I'm incorrectly making/not-making.

> If we can't trust what is running within our infrastructure, how can we trust the infrastructure managing JWT tokens and their authentication?

We have to root our trust somewhere, but we can confine it to a something that is extremely locked down and audited (such as the key infra). The strawman in the appendix addresses some of this, but let's say:
1. fetcher generates a keypair (private key stays in memory so it's lost on process exit)
2. fetcher registers the public key and provenance info (repo, entrypoint, commit, etc) with infra, which logs that data in a DB (we'll call this abstract thing the _keydb_) and returns a JWT bearer token for auth
3. manifests are signed with generated key
4. upload endpoint ensures JWT token is legit, validates the manifest's signature with the public key in the JWT bearer token

> I'd love to learn more and understand how this makes it safer, or if the goal is more being able to track exactly where data came from and what processed it.

I'm shooting for both.

When it comes to safety, the following scenarios can play out if we get owned:
* manifest/doc storage compromised:
  * action: attacker manages to get write access our manfiest/document store and changes a bunch of things
  * response: we'd verify the signatures of all manifests against what we have in the keydb and find anything that was inserted or tampered with (no solution for deletions)
* fetcher environment compromised:
  * action: attacker can run code in the fetcher environment to upload bad record data
  * response: we can identify the key used to upload the bad data and revoke the key in the keydb, causing verifications to fail
* fetcher's key material leaks:
  * action: a fetcher's private key is exfil'd and it's no longer actually private
  * response: revoke the key in the keydb, causing verifications to fail
* keydb is compromised
  * action: attacker can write arbitrarily to the keydb
  * response: if you can identify the tampering, revoke the relevant entries. If not, blow it away, restore from more trusted backups, re-fetch what you need to re-fetch.

As for tracking origin, this would allow us to confidently say "this data came from this scraper at this revision and this time", and then other things can build off of that (with processes which are ideally also verifiable).

Thoughts?